### PR TITLE
Remove build_helper

### PR DESCRIPTION
**Pull request by bjorn3**
The majority of the code is only used by either rustbuild or
rustc_llvm's build script. Rust_build is compiled once for rustbuild and
once for every stage. This means that the majority of the code in this
crate is needlessly compiled multiple times. By moving only the code
actually used by the respective crates to rustbuild and rustc_llvm's
build script, this needless duplicate compilation is avoided.

### CODE REVIEW

File: compiler/rustc_llvm/build.rs [hunk 1]
```diff
@@ -14,13 +14,74 @@ fn detect_llvm_link() -> (&'static str, &'static str) {
...
+        if entry.file_type().unwrap().is_dir() {
+            stack.extend(path.read_dir().unwrap().map(|e| e.unwrap()));
+        } else {
+            println!("cargo:rerun-if-changed={}", path.display());
+        }
+    }
+}
+
+#[track_caller]
+fn output(cmd: &mut Command) -> String {
```
Comment (bjorn3) [inline compiler/rustc_llvm/build.rs:60]: This is the only function used by both rustc_llvm's build script and by rustbuild.

**Comment by Mark-Simulacrum** [comment]
r=me with nit fixed

File: src/bootstrap/build_helper.rs [hunk 2]
```diff
@@ -1,9 +1,7 @@
-use std::ffi::{OsStr, OsString};
```
[the remark this answers is not in the document]
Comment (bjorn3) [reply src/bootstrap/build_helper.rs]: Done

**Comment by bjorn3** [comment]
Rebased

@bors r=Mark-Simulacrum

**Comment by matthiaskrgr** [comment]
@bors r- 
failed in a rollup  https://github.com/rust-lang/rust/pull/94606#issuecomment-1059293558

**Comment by bjorn3** [comment]
Fixed and rebased.

**Comment by Mark-Simulacrum** [comment]
r=me squashed in (2 commits is good)

**Comment by bjorn3** [comment]
Squashed

@bors r=Mark-Simulacrum

**Pull request merged**